### ISSUE TITLE
Bug in Gallery view related to metadata display

### ISSUE THREAD

**Issue by caaster** [projectblacklight/spotlight#1802]
There is a bug in gallery view related to metadata display. Here is how I am able to generate this bug.

I go to Dashboard > Configuration > Metadata and uncheck all metadata fields for Gallery view.

Then I go to Dashboard > Curation > Browse > Edit for a browse category, and check the radio button to designate Gallery as the default view.

Then, I go to my exhibit home page > Browse > click on Browse category I just configured, and I see all metadata displaying in the Gallery view. I expect to see only the default title displayed in Gallery view, but instead I see all metadata displayed in Gallery view.

There is some other condition that must be present for this bug to manifest that I have not effectively been able to sleuth, because Gary does not have the same experience as what I described above. But I have been able to repeat this bug/experience in Chrome, Firefox and Safari myself.

**Comment by starsplatter** [comment]
We see this locally as well. The reason not everyone sees it the same way is it is dependent on whether you have your main search results view set to list or not. If you set the main result to gallery and then go to browse, gallery works fine in browse.  If you set your main result to list and then try to look at browse in gallery, it is messed up. We're looking here to see if we can find a fix.

**Comment by drh-stanford** [comment]
I was successful in reproducing this problem on my development server

**Comment by drh-stanford** [comment]
So, I dug into the spotlight/blacklight code for a while and I wasn't successful finding where the dynamic configuration data was being drawn from. The Exhibit model has a `blacklight_configuration` relation that pulls the configuration data from the dadtabase, but I don't see where that's applied to the view_type. Any thoughts on this code path?

**Comment by drh-stanford** [comment]
Ok, so I'm unable to reproduce this problem now. I tried a new exhibit, new browse category and neither were able to reproduce the problem. It's contingent on some state for the exhibit as I was able to reproduce it yesterday.

**Comment by cbeer** [comment]
I assume this is related to caching -- I bet we don't bust the cache if metadata fields change, and I guess we need to?

**Comment by drh-stanford** [comment]
Which cache did you have in mind? 

I've tried to reproduce it several times and when I've been successful, if I click between List and Gallery view a couple times, the Gallery view fixes itself (as described in the second comment above).

I looked at the network loads and it's sending back a no-cache (`Cache-Control:max-age=0, private, must-revalidate`) response for the page. I don't understand enough about turbolinks but it appears that it's caching the view data and async loading it when you click on the Gallery icon.

**Comment by cbeer** [comment]
Yeah, I was thinking browser cache. If we're sending the right headers, then I dunno.

**Comment by drh-stanford** [comment]
We could try running `Turbolinks.clearCache()` in the browser JavaScript after we do a save of the metadata configuration.

**Comment by drh-stanford** [comment]
I have a branch that optionally disables turbolinks for the View controls here:

  https://github.com/projectblacklight/blacklight/compare/optionally-disable-turbolinks

I don't particularly like this solution but it might be something we can use to start a discussion on a better solution.

**Comment by drh-stanford** [comment]
@jkeck @cbeer I'm thinking that this issue is a "won't fix" as it is a turbolinks caching issue AFAICT, and it's not a problem that affects users browsing the exhibit. what are your thoughts on this? The options for fixing it are:

(a) run `Turbolinks.clearCache()` anytime we change the metadata display configuration state (but this does not fix the problem of an existing window open)
(b) disable turbolinks entirely on View switching (the https://github.com/projectblacklight/blacklight/commit/1958c8e2b31d1a9241bd9c38b9d2ddb2c483275d code)
(c) something else?

**Comment by jkeck** [comment]
Oh, so is this only caused by changing configurations with a current window/tab open (e.g. as a user, I would never see this unless somebody changed the configuration while I was actively toggling between views)?  And that a potential solution would be to simply refresh search results page once a configuration change has happened?

If that's the case, then I agree, I'm not sure there is much that we can do to fix this.  It seems like more of a bug that will most likely get unearthed by people testing for the bug, and not necessarily by real users in the course of using the application.

**Comment by drh-stanford** [comment]
I've gone through the data saved in the database many times and I have yet to see it corrupted. So, I don't think it's a data consistency problem (which was my initial assumption). The problem _looks_ like a data corruption issue but it corrects itself if you do a hard refresh in the browser. I think option (a) may mitigate the problem but it won't "fix" it in cases where they're using multiple windows/tabs.

**Comment by caaster** [comment]
@jkeck this has happened to me more than once in building an exhibit -- when I am trying to configure the exhibit as I need to. This is how I unearthed it. Sometime maybe the team can give me a little tutorial to help me understand better why this can't be fixed & what is happening. Lunch is on me for that!!

**Comment by aeschylus** [comment]
I was able to reproduce this error by coming into the gallery from the general "browse" tab. Selecting the browse category from the browse tab will render the gallery with the metadata displayed (even with all metadata checkboxes unchecked in the admin area). However, as noted above, clicking the list view and then clicking back to the gallery view using the settings buttons on the browse category view will cause the gallery to render without the metadata (as desired).

Jessy noticed that the url parameters in these two cases are different, and that may be to blame for the behaviour. When rendering the browse category after a click on its placeholder in the browse tab, the url is:
`https://exhibits-stage.stanford.edu/early-maps-demo/browse/all-exhibit-items`,
but when navigating from the top view toggle, the url is:
`https://exhibits-stage.stanford.edu/early-maps-demo/browse/all-exhibit-items?view=gallery`,
so it seems the query parameter is affecting the rendering by applying different settings.

Looking at the spotlight codebase, I think the place this is going awry. I'll find the other areas shortly (such as the other index view:

https://github.com/projectblacklight/spotlight/blob/master/app/views/spotlight/browse/show.html.erb#L24-L28

**Comment by aeschylus** [comment]
Isolated setup:

1. With a fresh pull from master and the default seed data loaded into spotlight, create two new browse categories from the results of searching "zuyd-zee" (which will include https://purl.stanford.edu/sn161bw2027). These items have metadata available that can be used for exhibiting the bug. 
1. In one browse category, set the default view type to "list". In the second category, do not set the default view type, but ensure that all metadata fields are unchecked in the metadata field selection tab for the gallery section in the "metadata" menu of the exhibit dashboard. Save settings.
1. In the default list view exhibit, navigating in from the browse section of spotlight, the "list" view is showing, so that configuration is picked up.
1. In the browse category with no view chosen, clicking into the category from the browse tab will load the gallery view, but metadata will be showing even though it is configured not to show. Logging the viewType, we get `nil`. The first branch of this logic is never being reached: https://github.com/projectblacklight/spotlight/blob/master/app/views/spotlight/browse/show.html.erb#L24-L28, so the question is how does blacklight provide this configuration (`document_index_view_type`), and how do we override it correctly or consistently in spotlight.
https://github.com/projectblacklight/blacklight/blob/f0e814b1b738bd25d3734039a298eb725663386f/app/helpers/blacklight/render_partials_helper_behavior.rb#L11
